From f390154dd87aec7ccee4f7f11ccb7c9d2be30bc4 Mon Sep 17 00:00:00 2001
From: Gao Nian <nian.gao@siemens.com>
Date: Thu, 2 Jan 2020 14:00:09 +0800
Subject: [PATCH 15/23] fix: rproc r5-0 set_config failed in linux

Signed-off-by: Gao Nian <nian.gao@siemens.com>
---
 drivers/remoteproc/ti_k3_r5_remoteproc.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/drivers/remoteproc/ti_k3_r5_remoteproc.c b/drivers/remoteproc/ti_k3_r5_remoteproc.c
index f9c2e474c4ea..8b9e9037ecd8 100644
--- a/drivers/remoteproc/ti_k3_r5_remoteproc.c
+++ b/drivers/remoteproc/ti_k3_r5_remoteproc.c
@@ -680,6 +680,7 @@ static int k3_r5_rproc_configure(struct k3_r5_rproc *kproc)
 	u32 ctrl = 0, cfg = 0, stat = 0;
 	u32 set_cfg = 0, clr_cfg = 0;
 	u64 boot_vec = 0;
+	u32 lockstep_en;
 	int ret;
 
 	core0 = list_first_entry(&cluster->cores, struct k3_r5_core, elem);
@@ -693,8 +694,8 @@ static int k3_r5_rproc_configure(struct k3_r5_rproc *kproc)
 	dev_dbg(dev, "boot_vector = 0x%llx, cfg = 0x%x ctrl = 0x%x stat = 0x%x\n",
 		boot_vec, cfg, ctrl, stat);
 
-	if (!(stat & PROC_BOOT_STATUS_FLAG_R5_LOCKSTEP_PERMITTED) &&
-	    cluster->mode) {
+	lockstep_en = stat & PROC_BOOT_STATUS_FLAG_R5_LOCKSTEP_PERMITTED;
+	if ((!lockstep_en) && cluster->mode) {
 		dev_err(cluster->dev, "lockstep mode not permitted, force configuring for split-mode\n");
 		cluster->mode = 0;
 	}
@@ -703,7 +704,8 @@ static int k3_r5_rproc_configure(struct k3_r5_rproc *kproc)
 	boot_vec = 0x0;
 	if (core == core0) {
 		clr_cfg = PROC_BOOT_CFG_FLAG_R5_TEINIT;
-		clr_cfg |= PROC_BOOT_CFG_FLAG_R5_LOCKSTEP;
+		if (lockstep_en)
+			clr_cfg |= PROC_BOOT_CFG_FLAG_R5_LOCKSTEP;
 	}
 
 	if (core->atcm_enable)
-- 
2.17.1

