From 5029d8c4dcef1dfee31b0bcafa779ed4c1194f50 Mon Sep 17 00:00:00 2001
From: Sheng Long Wang <shenglong.wang.ext@siemens.com>
Date: Mon, 25 Nov 2019 16:34:43 +0800
Subject: [PATCH] feat:add optee support to iot2050 platform

Signed-off-by: Sheng Long Wang <shenglong.wang.ext@siemens.com>
---
 core/arch/arm/plat-k3/platform_config.h | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/core/arch/arm/plat-k3/platform_config.h b/core/arch/arm/plat-k3/platform_config.h
index 186e7e69..893d4bf5 100644
--- a/core/arch/arm/plat-k3/platform_config.h
+++ b/core/arch/arm/plat-k3/platform_config.h
@@ -13,8 +13,13 @@
 #define UART1_BASE      0x02810000
 #define UART2_BASE      0x02820000
 
+#if defined(PLATFORM_FLAVOR_iot2050)
+/*MAIN_UART1*/
+#define CONSOLE_UART_BASE       UART1_BASE
+#else
 /* UART0 */
 #define CONSOLE_UART_BASE       UART0_BASE
+#endif
 #define CONSOLE_BAUDRATE        115200
 #define CONSOLE_UART_CLK_IN_HZ  48000000
 
-- 
2.22.0

