From 1c315cc40a232f12ddf7897732f268187fbea360 Mon Sep 17 00:00:00 2001
From: Gao Nian <nian.gao@siemens.com>
Date: Tue, 12 Nov 2019 19:26:20 +0800
Subject: [PATCH 06/15] mmc:change SD to mmc0 and EMMC-Flash to mmc1

Signed-off-by: Gao Nian <nian.gao@siemens.com>
---
 arch/arm/dts/iot2050-advanced.dts | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/arch/arm/dts/iot2050-advanced.dts b/arch/arm/dts/iot2050-advanced.dts
index 83040eb2b3..92703d2254 100644
--- a/arch/arm/dts/iot2050-advanced.dts
+++ b/arch/arm/dts/iot2050-advanced.dts
@@ -14,6 +14,11 @@
 	compatible =  "siemens,iot2050", "ti,am654";
 	model = "Siemens IOT2050 Advanced Base Board";
 
+	aliases {
+		mmc0 = &sdhci1;
+		mmc1 = &sdhci0;
+	};
+
 	memory@80000000 {
 		device_type = "memory";
 		/* 2G RAM */
-- 
2.17.1

