From 98cc8b6c0f116019e0b47a2c1a063d3dac5d5598 Mon Sep 17 00:00:00 2001
From: Gao Nian <nian.gao@siemens.com>
Date: Tue, 12 Nov 2019 19:30:14 +0800
Subject: [PATCH 07/15] feat:add scripts to select fdt name

Signed-off-by: Gao Nian <nian.gao@siemens.com>
---
 include/configs/iot2050.h | 22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/include/configs/iot2050.h b/include/configs/iot2050.h
index 81dbf39537..8f7b344cb3 100644
--- a/include/configs/iot2050.h
+++ b/include/configs/iot2050.h
@@ -79,6 +79,25 @@
     "uuid_compare=for uuid_other in ${uuids}; do if test ${uuid_conflict} = no && env exists uuid_get && test ${uuid_other} = ${uuid_get}; then  echo \"** ERROR: uuid conflicts: \"${uuid}\" !!! **\"; setenv uuid_conflict yes; fi; done\0"  \
     "uuid_verify=setenv uuid_conflict no; setenv uuid_break no; for target in ${boot_targets}; do if test ${uuid_break} = no; then env delete uuid_tmp; run uuid_get_${target}; setenv uuid_get ${uuid_tmp}; run uuid_list; run uuid_compare; if test ${target} = ${devtype}${devnum}; then setenv uuid_break yes; fi; fi; done\0"
 
+#define EXTRA_ENV_SELECT_FDT                           \
+    "fdt_found=no\0"                                  \
+    "fdt_select=if env exists board_name; then "       \
+                    "if test ${board_name} = IOT2050-BASIC; then "          \
+                        "env set fdtfile siemens/iot2050-basic.dtb;"        \
+                        "env set fdt_found yes;" \
+                    "else if test ${board_name} = IOT2050-ADVANCED; then "  \
+                        "env set fdtfile siemens/iot2050-advanced.dtb;"     \
+                        "env set fdt_found yes;" \
+                    "else "                                                  \
+                        "echo ** ERROR: can not find fdt file for ${board_name} **;" \
+                        "env set fdt_found no;" \
+                    "fi; fi;"     \
+                "else "           \
+                    "echo ** Warning: board_name is not defined, default boot as IOT2050-BASIC **;" \
+                    "env set fdtfile siemens/iot2050-basic.dtb;" \
+                    "env set fdt_found yes;" \
+                "fi;\0"
+
 /* U-Boot general configuration */
 #define EXTRA_ENV_IOT2050_BOARD_SETTINGS				\
 	"loadaddr=0x80080000\0"						\
@@ -114,7 +133,8 @@
 #define CONFIG_EXTRA_ENV_SETTINGS					\
 	BOOTENV								\
 	EXTRA_ENV_IOT2050_BOARD_SETTINGS	\
-	EXTRA_ENV_UUID_VERIFY
+	EXTRA_ENV_UUID_VERIFY   \
+	EXTRA_ENV_SELECT_FDT
 #endif
 
 #define CONFIG_SUPPORT_EMMC_BOOT
-- 
2.17.1

