From 6da565a50598aa74ab6ef1654350c30274166295 Mon Sep 17 00:00:00 2001
From: "le.jin@siemens.com" <le.jin@siemens.com>
Date: Sun, 16 Feb 2020 18:39:15 +0800
Subject: [PATCH 04/18] am65x:fix usb hub issue for disabling charge detect

Signed-off-by: le <le.jin@siemens.com>
---
 common/main.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/common/main.c b/common/main.c
index d926894f61..c36119be28 100644
--- a/common/main.c
+++ b/common/main.c
@@ -14,6 +14,7 @@
 #include <asm/io.h>
 #include <dm/uclass.h>
 #include <led.h>
+#include <asm/gpio.h>
 /*
  * Board-specific Platform code can reimplement show_boot_progress () if needed
  */
@@ -73,8 +74,10 @@ void main_loop(void)
 	if (cli_process_fdt(&s))
 		cli_secure_boot_cmd(s);
 
-	autoboot_command(s);
+	/* change CTRL_MMR register to let serdes0 not output USB3.0 signals. */
+	writel(0x3, 0x00104080);
 
+	autoboot_command(s);
 #if defined(CONFIG_TARGET_IOT2050_ADVANCED_A53) || defined(CONFIG_TARGET_IOT2050_BASIC_A53)
 	/* Turn on the LED when OS is not started  */
 	iot2050_show_set_light("redled",LEDST_ON);
-- 
2.17.1

