From bf4414b3d509a9497f8699e9101f7de8a06bdcae Mon Sep 17 00:00:00 2001
From: Su Bao Cheng <baocheng.su@siemens.com>
Date: Tue, 23 Apr 2019 10:07:17 +0800
Subject: [PATCH 2/8] Add support for U9300C TD-LTE module

Author: Gao Nian  <nian.gao@siemens.com>
Signed-off-by: Su Bao Cheng <baocheng.su@siemens.com>
---
 drivers/usb/serial/option.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/usb/serial/option.c b/drivers/usb/serial/option.c
index eb5538a44ee9..defcba6e0ed2 100644
--- a/drivers/usb/serial/option.c
+++ b/drivers/usb/serial/option.c
@@ -395,6 +395,9 @@ static void option_instat_callback(struct urb *urb);
 /* iBall 3.5G connect wireless modem */
 #define IBALL_3_5G_CONNECT			0x9605
 
+/* LONGSUNG U9300C TD-LTE wireless modem */
+#define LONGSUNG_U9300C   0x9b3c
+
 /* Zoom */
 #define ZOOM_PRODUCT_4597			0x9607
 
@@ -1857,6 +1860,7 @@ static const struct usb_device_id option_ids[] = {
 	  .driver_info = RSVD(4) },
 	{ USB_DEVICE(LONGCHEER_VENDOR_ID, ZOOM_PRODUCT_4597) },
 	{ USB_DEVICE(LONGCHEER_VENDOR_ID, IBALL_3_5G_CONNECT) },
+	{ USB_DEVICE(LONGCHEER_VENDOR_ID, LONGSUNG_U9300C) },
 	{ USB_DEVICE(HAIER_VENDOR_ID, HAIER_PRODUCT_CE100) },
 	{ USB_DEVICE_AND_INTERFACE_INFO(HAIER_VENDOR_ID, HAIER_PRODUCT_CE81B, 0xff, 0xff, 0xff) },
 	/* Pirelli  */
-- 
2.17.1

