From c163974fa09e95399a53f229da3632aeb4236b0e Mon Sep 17 00:00:00 2001
From: Sheng Long Wang <shenglong.wang.ext@siemens.com>
Date: Tue, 13 Aug 2019 09:30:38 +0800
Subject: [PATCH 4/8] fix: disable usb lpm to fix usb device reset

---
 drivers/usb/core/hub.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/usb/core/hub.c b/drivers/usb/core/hub.c
index 4d3de33885ff..e3e8bbf5ef8d 100644
--- a/drivers/usb/core/hub.c
+++ b/drivers/usb/core/hub.c
@@ -4494,7 +4494,7 @@ static void hub_set_initial_usb2_lpm_policy(struct usb_device *udev)
 	struct usb_hub *hub = usb_hub_to_struct_hub(udev->parent);
 	int connect_type = USB_PORT_CONNECT_TYPE_UNKNOWN;
 
-	if (!udev->usb2_hw_lpm_capable || !udev->bos)
+	//if (!udev->usb2_hw_lpm_capable || !udev->bos)
 		return;
 
 	if (hub)
-- 
2.17.1

