From 637d98fb1d5d1eb237d581330b30b89ee822caec Mon Sep 17 00:00:00 2001
From: zengchao <chao.zeng@siemens.com>
Date: Wed, 6 Nov 2019 11:21:49 +0800
Subject: [PATCH 5/8] setting the RJ45 port led behavior

Signed-off-by: zengchao <chao.zeng@siemens.com>
---
 drivers/net/phy/dp83867.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/net/phy/dp83867.c b/drivers/net/phy/dp83867.c
index 1b65c8a94e8c..3bdb3862a310 100644
--- a/drivers/net/phy/dp83867.c
+++ b/drivers/net/phy/dp83867.c
@@ -44,6 +44,10 @@
 #define DP83867_10M_SGMII_CFG   0x016F
 #define DP83867_10M_SGMII_RATE_ADAPT_MASK BIT(7)
 
+/*RJ45 led configuration*/
+#define DP83867_LEDCR_1      0x0018
+#define RJ45_LED_SETTING     0x665b
+
 #define DP83867_SW_RESET	BIT(15)
 #define DP83867_SW_RESTART	BIT(14)
 
@@ -463,6 +467,9 @@ static int dp83867_config_init(struct phy_device *phydev)
 			       mask, val);
 	}
 
+	/*Set the RJ45 led action*/
+	phy_write_mmd(phydev, DP83867_DEVADDR, DP83867_LEDCR_1, RJ45_LED_SETTING);
+
 	return 0;
 }
 
-- 
2.17.1

